Introduce ability to skip CRD installation with App/Chart Operator

yasn77 opened the pull request:
Towards giantswarm/giantswarm#13746

Add ability to define if the installation of CRD should be skipped.

## Checklist

- [x] Consider SIG UX feedback.
- [x] Update changelog in CHANGELOG.md.
- [x] If adding a new type, ensure that it is added to the [CRD unit tests](https://github.com/giantswarm/apiextensions/blob/d78aba91d578d56ef49f58e336035d35edc4e1b0/pkg/crd/crd_test.go#L9).

Commit:
Introduce SkipCRDInstall

Towards https://github.com/giantswarm/giantswarm/issues/13746

Add ability to define if the installation of CRD should be skipped.

Commit:
Update Changelog

Commit:
Fix tests

tomahawk28 commented on pkg/apis/application/v1alpha1/chart_types.go:
```diff
@@ -76,6 +76,8 @@ type ChartSpec struct {
 // +k8s:openapi-gen=true
 type ChartSpecConfig struct {
+	// Whether to install CRDs that might be supplied with the chart
+	SkipCRDInstall bool `json:"skipCrdInstall"`
```
I think `SkipCRD` is enough WDYT?

```suggestion
	SkipCRD bool `json:"skipCrdInstall"`
```

yasn77 replied:
I am going to go with Ross suggestion and add an `Install` struct. Thanks 👍🏽

marians replied:
Also see my comment https://github.com/giantswarm/apiextensions/pull/657/files#r558190783 which I think also applies here.

tomahawk28 reviewed (commented):
Please execute `make` from the top of the directory and commit auto-generated files.

rossf7 commented on docs/cr/application.giantswarm.io_v1alpha1_chart.yaml:
```diff
@@ -16,6 +16,7 @@ spec:
       name: f2def-chart-values
       namespace: f2def
       resourceVersion: ""
+    skipCrdInstall: false
```
```suggestion
  install:
    skipCRDs: false
```

@yasn77 So far we've done these type of config extensions by adding annotations like for skipping forced upgrades. But I think we should start extending the CRD which is fine as long as we don't make breaking changes.

I compared this with the helmrelease CRD as I was interested to see how Flux do this. I like how they structure this. As its aligned with the helm option.  WDYT?

https://toolkit.fluxcd.io/components/helm/helmreleases/

Also I think we need to make both install and skipCRDs optional. As in most cases we don't need to set this option.

yasn77 replied:
Yeh I like adding `Install.SkipCRDs`, it gives us an option of adding additional install time options such as skipping hooks (similar to fluxcd approach)

rossf7 commented on pkg/apis/application/v1alpha1/app_types.go:
```diff
@@ -85,6 +85,8 @@ type AppSpec struct {
 // +k8s:openapi-gen=true
 type AppSpecConfig struct {
+	// Whether to install CRDs that might be supplied with the chart
+	SkipCRDInstall bool `json:"skipCrdInstall"`
```
```suggestion
	SkipCRDInstall bool `json:"skipCrdInstall,omitempty"`
```

Adding omitempty will make the field optional as per my other comment.

yasn77 replied:
I am not sure about making it optional.... Doesn't a zero value boolean in go default to `false`? which is exactly what we want. It would conditional logic in the code similar right? something like...
```golang
if cr.Install.SkipCRDs {
...
}
```
Rather trying to test for the existence of the bool then test for true/false

rossf7 replied:
Yes in the operator the boolean will default to false. But one thing to watch is the install struct as it could be nil.

The omitempty is about keeping the app CRs cleaner when they are viewed in kubectl etc. So the setting will only appear if the user has requested CRDs to be skipped.

yasn77 commented:
> Please execute `make` from the top of the directory and commit auto-generated files.

Should this be a pre-commit step then?

Commit:
Use struct for install

Instead of single boolean, use an Install struct which
contains skipCRDs boolean

(suggested in PR https://github.com/giantswarm/apiextensions/pull/657)

Commit:
Add files from `make generate`

Commit:
Revert "Add files from `make generate`"

This reverts commit fd0ba859acb49d23af2235b9a418bd71a8534f76.

Commit:
make generate and go import

Commit:
Add note in README regarding make

Clarify what needs to be run

tomahawk28 commented on docs/cr/application.giantswarm.io_v1alpha1_chart.yaml:
```diff
@@ -16,6 +16,9 @@ spec:
       name: f2def-chart-values
       namespace: f2def
       resourceVersion: ""
+    skipCrdInstall: false
```
I think this should be removed. right?

yasn77 replied:
Yep... removing

Commit:
Remove unneeded value

pipo02mix commented on config/crd/v1/application.giantswarm.io_apps.yaml:
```diff
@@ -94,6 +94,17 @@ spec:
                     type: object
                 type: object
+              install:
+                description: Install is the config sed to deploy the app and is passed
```
I dont get the phrase, do you meant `set`?

pipo02mix replied:
```suggestion
                description: Install is the config used to deploy the app and is passed
                to Helm.
```

pipo02mix replied:
copy from other block ^^

yasn77 replied:
Yeh, was a typo.. Meant to be 'used'... Fixed now

Commit:
Fix typo

pipo02mix commented:
I tend to think just `crds`it is a better naming. Since it is under `install` property I would rather just use a name and not verb+name. Trying to keep consistency, WDYT?

rossf7 commented on pkg/apis/application/v1alpha1/app_types.go:
```diff
@@ -65,6 +65,10 @@ type AppSpec struct {
 	Config AppSpecConfig `json:"config,omitempty"`
+	// +kubebuilder:validation:Optional
+	// +nullable
+	// Install is the config used to deploy the app and is passed to chart operator.
```
```suggestion
	// Install is the config used when installing the app.
```

chart-operator is an implementation detail and may even change to the Flux helm-controller in the future. So can we drop that part?

rossf7 commented on pkg/apis/application/v1alpha1/app_types.go:
```diff
@@ -97,6 +101,14 @@ type AppSpecConfig struct {
+	// +kubebuilder:validation:Optional
+	// +nullable
+	// SkipCRDs decides whether to install CRDs that might be supplied with the chart
+	SkipCRDs bool `json:"skipCrds,omitempty"`
```
```suggestion
	SkipCRDs bool `json:"skipCRDs,omitempty"`
```

The JSON field should match the Go struct apart from the first char.

rossf7 reviewed (commented):
> I tend to think just crds it is a better naming. Since it is under install property I would rather just use a name and not verb+name.

@pipo02mix I also prefer keeping the CRs minimal but without the skip I think there isn't enough context. I also like that its aligned with the helm flag naming. WDYT?

```
helm install -h | grep skip-crds
      --skip-crds                    if set, no CRDs will be installed. By default, CRDs are installed if not already present
```

Commit:
Update pkg/apis/application/v1alpha1/app_types.go

Co-authored-by: Ross Fairbanks <rossf7@users.noreply.github.com>

Commit:
Update pkg/apis/application/v1alpha1/app_types.go

Co-authored-by: Ross Fairbanks <rossf7@users.noreply.github.com>

yasn77 commented:
> I tend to think just `crds`it is a better naming. Since it is under `install` property I would rather just use a name and not verb+name. Trying to keep consistency, WDYT?

@pipo02mix, In general, I prefer shorter things :) and yeh, it's a bit much when verb+name is already implied. However in this case, it matches with then what is actually passed to helm client.

I know that is a bit of an implementation detail, but we are not likely to move away from Helm any time soon and I don't expect Helm to drop `--skip-crd` in Helm 3. So I feel from a UX point of view, it's less of a cognitive load to use `skipCrd`, as the user (who most likely already is familiar with Helm) will know immediately what the functionality of that config will do.  It matches very closely with what they already know about Helm install.

marians commented:
@yasn77 As a reviewer, I'd find a description of the goal of this PR in the main title and comment helpful. Then I could look at the changes and check whether they make sense with respect to that goal.

yasn77 commented:
> @yasn77 As a reviewer, I'd find a description of the goal of this PR in the main title and comment helpful. Then I could look at the changes and check whether they make sense with respect to that goal.

Your right... My bad. I should have added more detail and not just used the branch name. I was experimenting with a new IDE to create the PR and missed what I usually take for granted.

marians commented on pkg/apis/application/v1alpha1/chart_types.go:
```diff
@@ -74,8 +78,18 @@ type ChartSpec struct {
+type ChartSpecInstall struct {
+	// +kubebuilder:validation:Optional
+	// +nullable
+	// SkipCRDs decides whether to install CRDs that might be supplied with the chart
+	SkipCRDs bool `json:"skipCrds,omitempty"`
```
There is a bit of a chance of a misinterpretation here. "skip" to me indicates that a CRD _will not_ be installed. However the description seems to indicate the opposite.

This could be resolved by making the description more explicit. Or by renaming the field.

Also it should be mentioned what the default is (if not set). This may be obvious in a Go context, but not in a CRD context.

```suggestion
	// SkipCRDs when true decides that CRDs which are supplied with the chart are not installed. Default: false.
	SkipCRDs bool `json:"skipCrds,omitempty"`
```

or

```suggestion
	// InstallCRDs decides that CRDs which are supplied with the chart will be installed (default). Set to false to omit CRD installation.
	InstallCRDs bool `json:"installCrds,omitempty"`
```

Commit:
Update pkg/apis/application/v1alpha1/chart_types.go

Co-authored-by: Marian Steinbach <marian@giantswarm.io>

Commit:
Fix tests

Commit:
Merge branch 'master' into add_skip_crd

rossf7 commented on pkg/apis/application/v1alpha1/app_types.go:
```diff
@@ -97,6 +101,14 @@ type AppSpecConfig struct {
+	// +kubebuilder:validation:Optional
+	// +nullable
+	// SkipCRDs when true decides that CRDs which are supplied with the chart are not installed. Default: false.
+	SkipCRDs bool `json:"skipCrds,omitempty"`
```
```suggestion
	SkipCRDs bool `json:"skipCRDs,omitempty"`
```

@yasn77 Could we make the JSON tag upper case as well? We don't have many abbreviations in CRDs but I think they're usually upper case.

rossf7 reviewed (approved):
LGTM just one query on the JSON tag

Commit:
Update pkg/apis/application/v1alpha1/app_types.go

Co-authored-by: Ross Fairbanks <rossf7@users.noreply.github.com>

Commit:
Capitalise JSON field

As suggested in https://github.com/giantswarm/apiextensions/pull/657#discussion_r559490779
